eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,122 @@
+  case class Watch(replyTo: ActorRef[Result], remoteNodeId: PublicKey, paymentHash: ByteVector32, timeout: BlockHeight) extends Command
+  private case class WrappedPeerReadyResult(result: PeerReadyNotifier.Result) extends Command
+  private case class WrappedCurrentBlockHeight(currentBlockHeight: CurrentBlockHeight) extends Command
+
+  sealed trait Result
+  case object AsyncPaymentTriggered extends Result
+  case object AsyncPaymentTimeout extends Result
+  // @formatter:on
+
+  def apply(): Behavior[Command] = Behaviors.setup { context =>
+    new AsyncPaymentTriggerer(context).initializing()
+  }
+}
+
+private class AsyncPaymentTriggerer(context: ActorContext[Command]) {
+  import AsyncPaymentTriggerer._
+
+  case class Watcher(replyTo: ActorRef[Result], timeout: BlockHeight, paymentHash: ByteVector32) {
+    def expired(currentBlockHeight: BlockHeight): Boolean = timeout <= currentBlockHeight
+  }
+  case class AsyncPaymentTrigger(notifier: ActorRef[PeerReadyNotifier.Command], watchers: Set[Watcher]) {
+    def update(currentBlockHeight: BlockHeight): Option[AsyncPaymentTrigger] = {
+      // notify watchers that timeout occurred before offline peer reconnected
+      val expiredWatchers = watchers.filter(_.expired(currentBlockHeight))
+      expiredWatchers.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+      // remove timed out watchers from set
+      val updatedWatchers: Set[Watcher] = watchers.removedAll(expiredWatchers)
+      if (updatedWatchers.isEmpty) {
+        // stop notifier for offline peer when all watchers time out
+        context.stop(notifier)
+        None
+      } else {
+        Some(AsyncPaymentTrigger(notifier, updatedWatchers))
+      }
+    }
+    def trigger(): Unit = watchers.foreach(e => e.replyTo ! AsyncPaymentTriggered)
+  }
+
+  private def initializing(): Behavior[Command] = {
+    Behaviors.receiveMessage[Command] {
+      case Start(switchboard) => watching(switchboard, Map())
+      case m => context.log.error(s"received unhandled message ${m.getClass.getSimpleName} before Start received.")
+        Behaviors.same
+    }
+  }
+
+  private def watching(switchboard: ActorRef[Switchboard.GetPeerInfo], triggers: Map[PublicKey, AsyncPaymentTrigger]): Behavior[Command] = {
+    val peerReadyResultAdapter = context.messageAdapter[PeerReadyNotifier.Result](WrappedPeerReadyResult)
+    context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[CurrentBlockHeight](WrappedCurrentBlockHeight))
+
+    Behaviors.receiveMessage[Command] {
+      case Watch(replyTo, remoteNodeId, paymentHash, timeout) =>
+        triggers.get(remoteNodeId) match {
+          case None =>
+            // add a new trigger
+            val notifier = context.spawn(Behaviors.supervise(PeerReadyNotifier(remoteNodeId, switchboard, Left(Duration.Inf)))
+              .onFailure(SupervisorStrategy.restart), s"peer-ready-notifier-$remoteNodeId-$timeout")
```

Comment:
I don't think the supervisor strategy should be restart, it is useless because the `PeerReadyNotifier` will be waiting for `NotifyWhenPeerReady` and won't actually be useful. I think we should instead:

- set the supervisor strategy to `stop` with an error log (this really shouldn't happen and indicates a bug)
- use `context.watchWith(notifier, CancelPeerPayments(remoteNodeId))` where `CancelPeerPayments` is a new `Command` that cancels the pending payments for that peer and removes it from the map

We could also do nothing and just wait for the payments to timeout, but it's a bit wasteful, we know they won't be triggered at that point so it's better if we cancel it to give some feedback to the sender.

Reply:
That makes sense; implemented in [839f406](https://github.com/ACINQ/eclair/pull/2491/commits/839f406a4ccf527c19b1419bddba9e6a808b2454). Instead of `CancelPeerPayments` I went with a private command called `NotifierStopped`. Do we also want a public `CancelPeerPayments` command to do the same thing, but without the error in the log?

Reply:
I'm still trying to understand the different supervision strategies.. if we used `resume` instead of `restart` it would resume in its previous state and resend/resubscribe to events to watch for the peer to be ready, but since this would be unexpected and not guaranteed to be cleanly handled by `PeerReadyNotifier` it's better to use `stop`?

Reply:
I'd like to add a simple test that verifies the probe receives a timeout when `AsyncPaymentTriggerer` receives the `NotifierStopped` message, but since the message is private,  I can't send it from the test framework. I also can't get a ref to the internally spawned `PeerReadyNotifier`. Any tips on how create a test for that?

Reply:
> Do we also want a public CancelPeerPayments command to do the same thing, but without the error in the log?

I think that's something we may add later when we implement the ability for the sender to cancel a pending async payments. It's a bit early to add it right now.

> I'd like to add a simple test that verifies the probe receives a timeout when AsyncPaymentTriggerer receives the NotifierStopped message, but since the message is private, I can't send it from the test framework.

Instead of making the message private, you can make it `private[relay]`, which will make it accessible from tests.
It is a bit leaky, but it makes testing much easier, so we've already used that trick several time across the codebase.
Since it's still marked `private[some-scope]` it is obvious to readers that this shouldn't normally be used, so it's somewhat ok.

Reply:
Thanks! I added the test in [026ea4e](https://github.com/ACINQ/eclair/pull/2491/commits/026ea4ed25c00b5c36da6f2b773a2004b69905d9)